### ISSUE TITLE
 Support data exchange between generic components (notebooks | scripts) and runtime specific components

### ISSUE THREAD

**Issue by ptitzler** [elyra-ai/elyra#1765]
**Is your feature request related to a problem? Please describe.**
Currently Elyra utilizes a dedicated object storage bucket to manage input and output artifacts for notebooks | scripts. A different approach is used for KFP/AA components, raising the need to provide an easy (and ideally transparent) mechanism to exchange data between notebooks | scripts and notebook components.

**Describe the solution you'd like**

See above

### Design
=================================================================================

This design is dependent on #1761 and the PR associated with it.

**Motivation**

Elyra now supports custom components ( components and operators native to the pipeline runtime), and is able to exchange data between those components respectively. However, the exchange of information is not possible between Elyra's own generic components (Notebook, python script and Rscript operations).  

**Considerations**

The changes proposed should bridge the gap between Elyra generic components and custom components using I/O concepts native to each runtime and avoid adding custom I/O code as much as possible. e.g. our S3 file handling in processors. 
 
**Kubeflow Pipelines**

As outlined in #1761, KFP uses the concept of input and output paths to handle the flow of information in a pipeline. This concept could be used in our generic components to achieve the same end results and at the same time streamline how we handle artifacts into one form/behavior

At the heart of the generic component is the use of the `bootstrapper.py` script which takes in a series values/arguments to handle pushing and pulling artifacts. This script can be baked into a component image and a component yaml created for each of the generic component types (notebook, python and r scripts). 

Generic components differ from custom components in that they will always have at least one, or more, local dependencies (the notebook or script, local dep files) that need to be pushed to the remote environment as part of the operation and pipeline. These files will need to be seeded in the pipeline for use by nodes in the pipeline. 

Given a simple pipeline:

`Generic component  A -> Custom component -> Generic component B`

Today, the kfp processor will gather all dependencies specified by generic components and pushes them into an S3 bucket for consumption by the pipeline's various steps. 
- The information needed to reference the needed files is recorded in each respective **generic** component step. 
- The `custom component` is unable to reference any of the local files pushed into the S3 bucket
- The `custom component` will not be able to reference any outputs from `generic component A`, nor provide outputs for consumption by `Generic component B`

Tasks:
- Refactoring all of Elyra's generic components to KFP components
  -  The number of outputs of generic components can vary but component yaml files are meant to be static in nature for reuse so look into building python components on the fly as a possible solution OR specify `outputpath` as a directory to keep the component static with an all or nothing approach( all outputs are placed into a single directory ).
- Create and include a seeding component as the first step of the pipeline, whenever a `generic component` is included in the pipeline
   - The kfp processor will still use the s3 client to upload dependencies
   - This seeding component will be responsible for pulling in ALL of the dependencies uploaded to S3 from the users local environment and making them available as outputs for the entire pipeline. See above for similar solution.
 
Expectations after completion:
- Custom components will be able to provide outputs for generic component consumption
- Generic components will be able to provide outputs for custom component consumption

**Apache Airflow**
Design in progress

**Comment by kevin-bates** [comment]
@ptitzler - could you please provide a link or more information regarding...
> A different approach is used for KFP components

Do you know if Airflow components have their own approach?

Also, are you suggesting that COS-related properties may be associated with component types (i.e., you could have a couple of flavors of COS within a pipeline)?  Or that we adjust our generic (built-in) components to honor other approaches such that a given pipeline only uses one form?

**Comment by ptitzler** [comment]
> more information regarding...

TL;DR:
Components declare their io requirements by declaring a parameter as 
 - `inputPath` -> KF passes a [generated] file name to the component implementation. The component implementation reads the file to access the actual input data. The fact that the data is made available as a file (as opposed to a data structure, e.g. string) is hidden from the user. 
 - `inputValue` -> KF passes a value to the component implementation. (suitable for small amounts of data)
 - `outputPath` -> KF passes a [generated] file name to the component implementation. The component implementation writes the output to the file.
 
Details:
- https://www.kubeflow.org/docs/components/pipelines/reference/component-spec/
- basic component example: https://github.com/ptitzler/kfp-component-tests/tree/main/download-file
- basic data exchange example between components: https://github.com/ptitzler/kfp-component-tests/blob/main/pipelines/create_and_run_pipeline-1.py (this is what https://github.com/elyra-ai/elyra/issues/1761 needs to implement)

> Do you know if Airflow components have their own approach?

I have not looked into that yet.

> Or that we adjust our generic (built-in) components to honor other approaches such that a given pipeline only uses one form?

This. I believe it'd be best if generic/Elyra components would behave like other components with respect to data exchange.

**Comment by ptitzler** [comment]
Elyra 3.8 introduced support for mounting data volumes in generic components, which makes it more efficient to exchange data between them. If we extend this capability to runtime-specific components users will be able to use them in combination with generic components, which isn't really feasible today because they use different data exchange mechanisms. @kevin-bates @kiersten-stokes

**Comment by ptitzler** [comment]
Data exchange is supported in Elyra 3.10+ via [mounted data volumes](https://elyra.readthedocs.io/en/latest/user_guide/pipelines.html?highlight=volumes#defining-pipeline-properties).
